src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -2789,8 +2859,13 @@ and solve_t' (env:Env.env) (problem:tprob) (wl:worklist) : solution =
       | Tm_app({n=Tm_uvar _}, _), Tm_uvar _
       | Tm_uvar _,                Tm_app({n=Tm_uvar _}, _)
       | Tm_app({n=Tm_uvar _}, _), Tm_app({n=Tm_uvar _}, _) ->
-        let f1, wl = destruct_flex_t t1 wl in
-        let f2, wl = destruct_flex_t t2 wl in
+        (* We need to do both ensure_no_uvar_subst calls before
+         * destructing since we might have the same uvar on
+         * both sides. See #1616. *)
+        let t1, wl = ensure_no_uvar_subst t1 wl in
+        let t2, wl = ensure_no_uvar_subst t2 wl in
+        let f1 = destruct_flex_t' t1 in
+        let f2 = destruct_flex_t' t2 in
```

Comment:
Why not retain the original calls, since those are now safely calling ensure_no_uvar_subst?

Reply:
It's required that we call `ensure_no_uvar_subst` on both sides before we destruct, since we may be dealing with the same `ctx_uvar` on each side. If we don't, the uvar returned by the first call to `destruct_flex_t` may be solved during the second call.

If the uvars carry the same substitution (like in the original issue) then there is no problem since the `ensure_no_uvar_subst` will now be a no-op, but if the substs are different we need to do it like this.

(I did not actually test that it's needed though, since I'm not sure how to even construct such an example, but I'm pretty sure it would explode again.)

Reply:
Ah, that's subtle. I didn't fully appreciate that from the comment. Maybe some kind of bold warning there with some version of your comment above copied (rather than just linked).